File: Java/lib/FormalATMj/src/gov/nasa/larcfm/ACCoRD/WCVTable.java
/*
 * Copyright (c) 2012-2016 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */
package gov.nasa.larcfm.ACCoRD;

import gov.nasa.larcfm.Util.ParameterData;
import gov.nasa.larcfm.Util.Units;
import gov.nasa.larcfm.Util.f;

public class WCVTable implements ParameterTable {
  protected double DTHR; // Distance threshold
  protected double ZTHR; // Vertical threshold
  protected double TTHR; // Time threshold
  protected double TCOA; // Time to co-altitude threshold

  /**
   * WCV table SARP concept
   */
  public WCVTable() {
    DTHR = Units.from("nmi",0.66);   
    ZTHR = Units.from("ft",450);
    TTHR = 35; // [s]
    TCOA = 0;  // [s]
  }

  /**
   * Copy constructor -- returns a fresh copy
   * @param tab
   */
  public WCVTable(WCVTable tab) {
    copyValues(tab);
  }
  
  /**
   * Table containing specified values (internal units)
   */
  
  public WCVTable(double dthr, double zthr, double tthr, double tcoa) {
    DTHR = dthr;
    ZTHR = zthr;
    TTHR = tthr;
    TCOA = tcoa;
  }

  public WCVTable copy() {
    return new WCVTable(this);
  }

  /**
   * Initialize this table's values with those of another table
   */
  public void copyValues(WCVTable tab) {
    DTHR = tab.DTHR;
    ZTHR = tab.ZTHR;
    TTHR = tab.TTHR;
    TCOA = tab.TCOA;
  }
  
  /**
   * Return horizontal distance threshold DTHR in internal units
   */

  public double getDTHR()  {
    return DTHR;
  }

  /**
   * Return horizontal distance threshold DTHR in u units
   */

  public double getDTHR(String u)  {
    return Units.to(u,DTHR);
  }

  /**
   * Return vertical distance threshold ZTHR in internal units
   */

  public double getZTHR()  {
    return ZTHR;
  }

  /**
   * Return vertical distance threshold ZTHR in u units
   */

  public double getZTHR(String u)  {
    return Units.to(u,ZTHR);
  }

  /**
   * Return horizontal time threshold TTHR in seconds
   */

  public double getTTHR()  {
    return TTHR;
  }
  
  public double getTTHR(String u)  {
    return Units.to(u, TTHR);
  }
  
  /**
   * Return vertical time threshold TCOA in seconds
   */

  public double getTCOA()  {
    return TCOA;
  }
  public double getTCOA(String u)  {
    return Units.to(u,TCOA);
  }
  
  /* 
   * Set horizontal distance threshold DTHR to val, which is specified in internal units
   */
  
  public void setDTHR(double val) {
    DTHR = val;
  }

  /*
   *  Set horizontal distance threshold DTHR to val, which is specified in u units
   */
  
  public void setDTHR(double val, String u) {
    DTHR = Units.from(u,val);
  }

  /* 
   * Set vertical distance threshold ZTHR to val, which is specified in u units
   */
  
  public void setZTHR(double val) {
    ZTHR = val;
  }

  /*
   *  Set vertical distance threshold ZTHR to val, which is specified in internal units
   */
  
  public void setZTHR(double val, String u) {
    ZTHR = Units.from(u,val);
  }

  /*
   *  Set horizontal time threshold TTHR to val, which is specified in seconds
   */
 
  public void setTTHR(double val) {
    TTHR = val;
  }

  public void setTTHR(double val, String u) {
    TTHR = Units.from(u,val);
  }
  
  
  /* 
   * Set vertical time threshold TCOA to val, which is specified in seconds
   */
  
  public void setTCOA(double val) {
    TCOA = val;
  }
  
  public void setTCOA(double val, String u) {
    TCOA = Units.from(u,val);
  }

  public ParameterData getParameters() {
    ParameterData p = new ParameterData();
    updateParameterData(p);
    return p;
  }
  
  public void updateParameterData(ParameterData p) {
    p.setInternal("WCV_DTHR",DTHR,"ft");
    p.setInternal("WCV_ZTHR",ZTHR,"ft");
    p.setInternal("WCV_TTHR",TTHR,"s");
    p.setInternal("WCV_TCOA",TCOA,"s");
  }

  public void setParameters(ParameterData p) {
    if (p.contains("WCV_DTHR")) {
      DTHR = p.getValue("WCV_DTHR");
    }
    if (p.contains("WCV_ZTHR")) {
      ZTHR = p.getValue("WCV_ZTHR");
    }
    if (p.contains("WCV_TTHR")) {
      TTHR = p.getValue("WCV_TTHR");
    }
    if (p.contains("WCV_TCOA")) {
      TCOA = p.getValue("WCV_TCOA");
    }
  }
  
  public String toString() {
    return "DTHR: "+Units.str("NM",DTHR)+"; ZTHR: "+Units.str("ft",ZTHR)+"; TTHR: "+
        Units.str("s",TTHR)+"; TCOA: "+Units.str("s",TCOA);
  }
  
  public String toPVS(int prec) {
    return "(# DTHR := "+f.FmPrecision(DTHR,prec)+", ZTHR := "+f.FmPrecision(ZTHR,prec)+
        ", TTHR := "+f.Fm1(TTHR)+", TCOA := "+f.Fm1(TCOA)+" #)";
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    long temp;
    temp = Double.doubleToLongBits(DTHR);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(TCOA);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(TTHR);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    temp = Double.doubleToLongBits(ZTHR);
    result = prime * result + (int) (temp ^ (temp >>> 32));
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    WCVTable other = (WCVTable) obj;
    if (Double.doubleToLongBits(DTHR) != Double.doubleToLongBits(other.DTHR))
      return false;
    if (Double.doubleToLongBits(TCOA) != Double.doubleToLongBits(other.TCOA))
      return false;
    if (Double.doubleToLongBits(TTHR) != Double.doubleToLongBits(other.TTHR))
      return false;
    if (Double.doubleToLongBits(ZTHR) != Double.doubleToLongBits(other.ZTHR))
      return false;
    return true;
  }
  
  public boolean contains(WCVTable tab) {
    return DTHR >= tab.DTHR && TCOA >= tab.TCOA && TTHR >= tab.TTHR && ZTHR >= tab.ZTHR;
  }
  
}

File: Modules/ACCoRD/src/WCVTable.cpp
/*
 * Copyright (c) 2012-2018 United States Government as represented by
 * the National Aeronautics and Space Administration.  No copyright
 * is claimed in the United States under Title 17, U.S.Code. All Other
 * Rights Reserved.
 */

#include "Units.h"
#include "WCVTable.h"
#include "format.h"
#include "ParameterData.h"

namespace larcfm {

WCVTable::WCVTable() {
  DTHR = Units::from("nmi",0.66);
  ZTHR = Units::from("ft",450);
  TTHR = 35; // [s]
  TCOA = 0;  // [s]
  units_["WCV_DTHR"] = "nmi";
  units_["WCV_ZTHR"] = "ft";
  units_["WCV_TTHR"] = "s";
  units_["WCV_TCOA"] = "s";
}

WCVTable::WCVTable(double dthr, double zthr, double tthr, double tcoa) {
  DTHR = std::abs(dthr);
  ZTHR = std::abs(zthr);
  TTHR = std::abs(tthr);
  TCOA = std::abs(tcoa);
  units_["WCV_DTHR"] = "m";
  units_["WCV_ZTHR"] = "m";
  units_["WCV_TTHR"] = "s";
  units_["WCV_TCOA"] = "s";
}

/**
 * Table containing specified values (specified units)
 */
WCVTable::WCVTable(double dthr, const std::string& udthr, double zthr, const std::string& uzthr,
    double tthr, const std::string& utthr, double tcoa, const std::string& utcoa) {
  DTHR = Units::from(udthr,std::abs(dthr));
  ZTHR = Units::from(uzthr,std::abs(zthr));
  TTHR = Units::from(utthr,std::abs(tthr));
  TCOA = Units::from(utcoa,std::abs(tcoa));
  units_["WCV_DTHR"] = udthr;
  units_["WCV_ZTHR"] = uzthr;
  units_["WCV_TTHR"] = utthr;
  units_["WCV_TCOA"] = utcoa;
}

WCVTable::~WCVTable() { }

double WCVTable::getDTHR() const {
  return DTHR;
}
double WCVTable::getDTHR(const std::string& u) const {
  return Units::to(u,DTHR);
}

double WCVTable::getZTHR() const {
  return ZTHR;
}
double WCVTable::getZTHR(const std::string& u) const {
  return Units::to(u,ZTHR);
}

double WCVTable::getTTHR() const {
  return TTHR;
}
double WCVTable::getTTHR(const std::string& u) const {
  return Units::to(u,TTHR);
}

double WCVTable::getTCOA() const {
  return TCOA;
}
double WCVTable::getTCOA(const std::string& u) const {
  return Units::to(u,TCOA);
}

void WCVTable::setDTHR(double val) {
  DTHR = std::abs(val);
}

void WCVTable::setDTHR(double val, const std::string& u) {
  setDTHR(Units::from(u,val));
  units_["WCV_DTHR"] = u;
}

void WCVTable::setZTHR(double val) {
  ZTHR = std::abs(val);
}
void WCVTable::setZTHR(double val, const std::string& u) {
  setZTHR(Units::from(u,val));
  units_["WCV_ZTHR"] = u;
}

void WCVTable::setTTHR(double val) {
  TTHR =std::abs(val);
}
void WCVTable::setTTHR(double val, const std::string& u) {
  setTTHR(Units::from(u,val));
  units_["WCV_TTHR"] = u;
}

void WCVTable::setTCOA(double val) {
  TCOA = std::abs(val);
}
void WCVTable::setTCOA(double val, const std::string& u) {
  setTCOA(Units::from(u,val));
  units_["WCV_TCOA"] = u;
}

ParameterData WCVTable::getParameters() const {
  ParameterData p;
  updateParameterData(p);
  return p;
}

void WCVTable::updateParameterData(ParameterData& p) const {
  p.setInternal("WCV_DTHR",DTHR,getUnits("WCV_DTHR"));
  p.setInternal("WCV_ZTHR",ZTHR,getUnits("WCV_ZTHR"));
  p.setInternal("WCV_TTHR",TTHR,getUnits("WCV_TTHR"));
  p.setInternal("WCV_TCOA",TCOA,getUnits("WCV_TCOA"));
}

void WCVTable::setParameters(const ParameterData& p) {
  if (p.contains("WCV_DTHR")) {
    setDTHR(p.getValue("WCV_DTHR"));
    units_["WCV_DTHR"] = p.getUnit("WCV_DTHR");
  }
  if (p.contains("WCV_ZTHR")) {
    setZTHR(p.getValue("WCV_ZTHR"));
    units_["WCV_ZTHR"] = p.getUnit("WCV_ZTHR");
  }
  if (p.contains("WCV_TTHR")) {
    setTTHR(p.getValue("WCV_TTHR"));
    units_["WCV_TTHR"] = p.getUnit("WCV_TTHR");
  }
  if (p.contains("WCV_TCOA")) {
    setTCOA(p.getValue("WCV_TCOA"));
    units_["WCV_TCOA"] = p.getUnit("WCV_TCOA");
  }
}

std::string WCVTable::getUnits(const std::string& key) const {
  std::map<std::string,std::string>::const_iterator got = units_.find(key);
  if (got == units_.end()) {
    return "unspecified";
  }
  return got->second;
}

bool WCVTable::equals(const WCVTable& t) const {
  if (DTHR != t.DTHR) return false;
  if (ZTHR != t.ZTHR) return false;
  if (TTHR != t.TTHR) return false;
  if (TCOA != t.TCOA) return false;
  return true;
}

std::string WCVTable::toString() const {
  return "WCV_DTHR = "+Units::str(getUnits("WCV_DTHR"),DTHR)+
      ", WCV_ZTHR = "+Units::str(getUnits("WCV_ZTHR"),ZTHR)+
      ", WCV_TTHR = "+Units::str(getUnits("WCV_TTHR"),TTHR)+
      ", WCV_TCOA = "+Units::str(getUnits("WCV_TCOA"),TCOA);
}

std::string WCVTable::toPVS() const {
  return "(# DTHR := "+FmPrecision(DTHR)+", ZTHR := "+FmPrecision(ZTHR)+
      ", TTHR := "+FmPrecision(TTHR)+", TCOA := "+FmPrecision(TCOA)+" #)";
}

bool WCVTable::contains(const WCVTable& tab) const {
  return DTHR >= tab.DTHR && ZTHR >= tab.ZTHR && TTHR >= tab.TTHR && TCOA >= tab.TCOA;
}

}
